dynamic host volumes: plugin specification doc (#24924)
hashicorp/nomad: website/content/docs/concepts/plugins/storage/host-volumes.mdx

    if [ ! -f "$path.ext4" ]; then
      # dd only writes to stderr, so safe to run without redirection
      dd if=/dev/zero of="$path.ext4" bs=1M count="$megs"
      # mkfs includes stdout, so we need to redirect that to stderr
      mkfs.ext4 "$path.ext4" 1>&2
    fi
    if ! is_mounted "$path"; then
      mount "$path.ext4" "$path"
    fi
}

delete_volume() {
  local path="$1"
  is_mounted "$path" && umount "$path"
  rm -rf "$path"
  rm -f "$path.ext4"
}

case "$1" in
  "fingerprint")
    printf '{"version": "%s"}' "$version"
    ;;
  "create")
    path="$(get_path)"
    stderr "creating volume at $path"
    create_volume "$path" "$DHV_CAPACITY_MIN_BYTES"
    # output what Nomad expects
    bytes="$(stat --format='%s' "$path.ext4")"
    printf '{"path": "%s", "bytes": %s}' "$path" "$bytes"
    ;;
  "delete")
    path="$(get_path)"
    stderr "deleting volume at $path"
    delete_volume "$path" ;;
  *)
    echo "unknown operation: $1"
    exit 1 ;;
esac
```

</Tab>
</Tabs>

## Specification

A plugin is registered with Nomad if it:

* Is an executable file (a script or binary)
* Is located in the [`client.host_volume_plugin_dir`][plugin_dir] directory

> * Responds appropriately to a `fingerprint` call

Reviewer: Just to make it unmistakable, maybe we should note that the plugins need to be installed on clients and not servers. That's "obvious", and certainly implies by the config value, but...

### Operations

To fully manage the lifecycle of a volume, plugins must fulfill all of the
following operations:

* [fingerprint](#fingerprint)
* [create](#create)
* [delete](#delete)

Nomad passes the operation as the first positional argument to the plugin.
That and other information are passed as environment variables. Environment
variables are prefixed with `"DHV_"` (i.e. Dynamic Host Volume).

Some variables may be required for certain plugins (namely, `DHV_CAPACITY_*`
for plugins that can restrict size). Most are for plugin author convenience.

#### fingerprint
<blockquote style={{borderLeft: "solid 1px #00ca8e"}}>

Nomad calls `fingerprint` when the client agent starts (or is reloaded with a
SIGHUP) to discover valid plugins. The returned "version" is used to register
the plugin on the Nomad node for volume scheduling.

**CLI arguments:** `$1=fingerprint`

**Environment variables:**

```
DHV_OPERATION=fingerprint
```

**Expected stdout:**

```
{"version": "0.0.1"}
```

**Requirements:**

* Must complete within 5 seconds, or Nomad will kill it

>   * should be much faster, as no actual work should be done

Reviewer: Maybe this sentence fragment should be part of the bullet above? (Or in the "considerations" section?)

* "version" value must be valid per the [hashicorp/go-version][go-version]
  golang package

</blockquote>

#### create
<blockquote style={{borderLeft: "solid 1px #00ca8e"}}>

Nomad calls `create` when you run [`nomad volume create`][cli-create] CLI or
use the [create API][api-create]).

You can run `create` again for the same volume if you include the volume
`id` in the volume specification.

Nomad also calls `create` when the agent starts, for each volume that was
previously created on the node, so plugins can ensure the volumes are available
after an agent restart or host reboot.

**CLI Arguments:** `$1=create`

**Environment variables:**

```
DHV_OPERATION=create

> DHV_HOST_PATH=/suggested/path/to/place/volume

Reviewer: Do we want to define this further? i.e. is this path the path of the directory where the plugin should create the volume, or the path of the volume itself?

Author: in `main` currently it is the latter, but we're changing it to be the former, i.e. it will be a directory wherein the plugin should create a new volume directory (or mount, or what have you)

which of these would be clearer?

A) different wording in the example path

```suggestion
DHV_HOST_PATH=/put/your/volume/in/this/directory
```

B) use the `{}` convention

```suggestion
DHV_HOST_PATH={directory to put the volume in}
```

C) a more realistic example

```suggestion
DHV_HOST_PATH=/opt/nomad/data/host_volumes
```

D) something else?

Reviewer: I like B here. That way we're not specifying anything that we don't intend to (ex. a relationship between the host path and the datadir).

DHV_NODE_ID={Nomad node ID}
DHV_NODE_POOL={Nomad node pool}
DHV_NAMESPACE={volume namespace}
DHV_VOLUME_NAME={name from the volume specification}
DHV_VOLUME_ID={volume ID assigned by Nomad}

> DHV_CAPACITY_MIN_BYTES={capacity_min from the volume spec}
> DHV_CAPACITY_MAX_BYTES={capacity_max from the volume spec}

Reviewer: The spec has the human-friendly values, so:

```suggestion
DHV_CAPACITY_MIN_BYTES={capacity_min from the volume spec, expressed in bytes}
DHV_CAPACITY_MAX_BYTES={capacity_max from the volume spec, expressed in bytes}
```

DHV_PARAMETERS={json of parameters from the volume spec, or: null}
DHV_PLUGIN_DIR={path to directory containing plugins}
```

**Expected stdout:**

```
{"path": "/path/to/created/volume", "bytes": 50000000}
```

**Requirements:**

* Must complete within 60 seconds, or Nomad will kill it

> * Must be idempotent - repeat runs produce the same result

Reviewer: Maybe explicitly say "for the same volume ID" and/or "for the same arguments"? It's not just that we're idempotent but that we'll update the volume too.

* Must be safe to run concurrently, per volume name per node
* Returned "path" must be the location of the created volume.
  Nomad will mount this path into workloads that request the volume.

</blockquote>

#### delete
<blockquote style={{borderLeft: "solid 1px #00ca8e"}}>

Nomad calls `delete` when you run [`nomad volume delete`][cli-delete] CLI or
use the [delete API][api-delete]).

Nomad also runs `delete` when an initial volume `create` operation
fails, since it may have been only partially completed.

**CLI Arguments:** `$1=delete`

**Environment variables:**

```
DHV_OPERATION=delete
DHV_HOST_PATH=/suggested/path/to/place/volumes
DHV_NODE_ID={Nomad node ID}
DHV_NODE_POOL={Nomad node pool}
DHV_NAMESPACE={volume namespace}
DHV_VOLUME_NAME={name from the volume specification}
DHV_VOLUME_ID={volume ID assigned by Nomad}
DHV_PARAMETERS={json of parameters from the volume spec, or: null}
DHV_PLUGIN_DIR={path to directory containing plugins}
```

**Expected stdout:** none (stdout is discarded)

**Requirements:**

* Must complete within 60 seconds, or Nomad will kill it

> * Must be idempotent - repeat runs produce the same result

Reviewer: It's implied by idempotency, but we should probably refine this further to explicitly say that calling delete on an already deleted volume must not return an error. Just so there are no doubts for authors.

> * Will be run after a partial (failed) create operation TODO TODO TODO !!! how about create=UPDATE ???

Reviewer: TODO!!!

Author: will fix in a follow-up PR :sweat_smile:

* Only one create/delete operation at a time is executed per volume `name`
  per node, and similarly by `id` on Nomad servers, but many create/delete
  operations for different volume IDs may run concurrently, even on the same
  node.
* Plugins that write into network storage need to take care not to delete
  remote/shared state by `name`, unless they know that there are no other
  volumes with workloads using that name.

### Configuration

* Per-_volume_ configuration should be set in the volume specification file's
  `parameters`. Per-_node_ configuration should be put in config file(s) as
  described next.
* There is no mechanism built into Nomad for plugin configuration. As a
  convention, we suggest placing any necessary configuration file(s) next to
  the executable plugin in the plugin directory. You may use `DHV_PLUGIN_DIR`
  to refer to the directory.
* If a plugin needs to retain state across operations (e.g. delete needs
  some value that was generated during create), then you may store that on
  the host filesystem, or some external data store of your choosing, perhaps
  even Nomad variables.

> [dhv]: TODO(1.10.0)?

Reviewer: The top-level DHV documentation is currently just the volume spec page. That holds most of the conceptual info.

Author: should this later point to `.../stateful-workloads#dynamic-host-volumes` (or similar) whenever that's written up?

the volume spec is very important, but doesn't quite clearly describe the holistic feature near the top, as it's more of a reference, so I hesitate to link it as the main source of high-level info. of course I do link to it when referring directly to the volume spec.

Reviewer: > should this later point to .../stateful-workloads#dynamic-host-volumes (or similar) whenever that's written up?

Probably, yeah. For CSI it's _this_ document in Concepts > Plugins that is the top-level conceptual document. So there's a bit of divergence here because in CSI we're not responsible for defining the plugin spec at all.
